DevOps task formal PR

LePaVuki opened the pull request:
DevOps task formal PR for the code review

The project consists of:
- Base repository with Angular and Node
- Dockerfile to set up an image of Nginx
- Nginx configuration file to serve an Angular app
- GitHub workflow, which runs following job on every push/PR to master:
  - Installs/Tests/Builds the Angular app
  - Builds a docker image based on the Angular app served by Nginx
  - Pushes created image to the GitHub registry

LePaVuki committed:
Create main.yml

Initial test of GitHub actions

LePaVuki committed:
Update main.yml

changed ti install, test

LePaVuki committed:
Update main.yml

trying with cd command

LePaVuki committed:
Update main.yml

attempt with npm ci and actions/setup-node@v3

LePaVuki committed:
Update main.yml

attempt with "working-directory" parameters
running:
- npm run start
- npm run test

LePaVuki committed:
changed readme

LePaVuki committed:
Update main.yml

fixed git fail

LePaVuki committed:
Made actions run tests for node locally

LePaVuki committed:
changed readme back

LePaVuki committed:
renamed workflow

LePaVuki committed:
Added ng build

LePaVuki committed:
Added docker image building to ci

LePaVuki committed:
added loging to Container registry and pushing docker image

LePaVuki committed:
changed logging username

LePaVuki committed:
added tags to the building step

LePaVuki committed:
Changed labeling and authorisation on push

LePaVuki committed:
renamed tasks within the job to test successful workflow run and email notification

LePaVuki committed:
Changed compose.yaml to pull the image from GitHub

LePaVuki committed:
changed user loggin in the Dockerfile and made compose pull from the GitHub

LePaVuki committed:
fixed user creation in dockerfile

LePaVuki committed:
chanfed directory of nginx.conf to copy to inside the container

LePaVuki committed:
version with 2 .conf files being copied both to ./angular and /etc/nginx/

LePaVuki committed:
disabled user creation in dockerfile

LePaVuki committed:
changed username in nginx.conf to Nginx and restored user creation in the dockerfile

LePaVuki committed:
changed to unpriveleged version of Nginx in Dockerfile

LePaVuki committed:
Fixed the name of the Nginx image in dockerfile

LePaVuki committed:
restored the username in nginx.conf to www-data, changed which nginx.conf file being copied inside the container

LePaVuki committed:
changed nginx image to pull a regular one

LePaVuki committed:
Changed Dockerfile and nginx.conf

Changed dockerfile to give nginx user privileges and deleted user mode from nginx.conf

LePaVuki committed:
Changed Dockerfile to copy angular-starter into Nginx folder at var/www/html/

LePaVuki committed:
chenged nginx.conf to only listen to 80 port (no ipv6)

LePaVuki committed:
Changed paths in Dockerfile to relative

LePaVuki committed:
Changed nginx.conf

Changed location in the server parameters to /index.html

LePaVuki committed:
Changed port forwarding in docker compose and root in nginx.conf

LePaVuki committed:
Final commit before pull

Modified .ignore files, since was building project locally for testing
Modified compose for correct port forwarding

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+  create-and-push-image:
+    runs-on: ubuntu-latest
+
+    permissions:
```
Is it really needed? Pls leave only the stuff without which pipeline doesn't work

LePaVuki replied:
Deleted unnecessary permissions

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+      uses: actions/checkout@v3
+
+    - name: INSTALL Node
+      working-directory: ./node
```
Should be `angular` dir instead of node

LePaVuki replied:
Changed all ./node working directories to ./angular

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+      id-token: write
+
+    steps:
+    - name: CHECKOUT
```
Let's define task steps in lower case, i.e. `Checkout` or `Install node` and etc. Ususally upper case is used for env variables or secrets

LePaVuki replied:
Changed names to lowercase

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+
+    - name: INSTALL Node
+      working-directory: ./node
+      run: npm install
```
Usually it is better to use `npm ci` - https://builtin.com/software-engineering-perspectives/npm-ci-vs-npm-install

LePaVuki replied:
Switched to npm ci instead

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+      run: npm install
+
+    - name: RUN SuperTest API Testing
+      working-directory: ./node
```
Let's specify default `working-directory` for a whole job to DRY

LePaVuki replied:
Specified the working directory: ./angular for every run command inside the whole job. Or should I have done something else? What is DRY? Env variable?

kseniyashaydurova replied:
It is a `Don't Repeat Yourself` principal in programming - https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

So using default `working dir` you don't need to do `cd angular` in each step

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+      working-directory: ./angular
+      run: |
+        npm link @angular/cli
+        ng build
```
Let's use native `build` command from package.json

LePaVuki replied:
Switched to npm run build

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+      with:
+        registry: ${{ env.REGISTRY }}
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
```
Let's use Github Actions automatic variables?
https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication
![image](https://github.com/user-attachments/assets/03746fff-1ba7-4b7d-9bfa-1a9df02c224d)

LePaVuki replied:
Changed to use of github.token in the workflow

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,61 @@
+        context: .
+        push: true
+        tags: ${{ steps.meta.outputs.tags }}
+        labels: ${{ steps.meta.outputs.labels }}
```
Don't see a step to send email notification on build result

LePaVuki replied:
The project currently uses aforementioned GitHub default email notification system. If it's not enough, please tell me

dmitry-mightydevops replied:
yes pls create a custom email handler as part of the workflow

kseniyashaydurova commented on .github/workflows/ci.yml:
Can you add a step to measure the time for execution of your entire pipeline and send that info in the email notification too?

LePaVuki replied:
Email notifications are being handled by GitHub by default for the user, pushing commit. After changing settings to notify about any outcome, I can receive an email notification while pushing to origin/main. Are those enough, or do I need to set up additional steps inside the workflow? And if I do, is there any actions from the marketplace specifically I should use?

An example of such email notifications, provided by GitHub, would be these emails I got:

---
## Success
![Success](https://github.com/user-attachments/assets/3c683ede-9614-4a77-ae37-0e55476e6226)
---
## Failure
![Failure](https://github.com/user-attachments/assets/a9f8cd48-fd75-4663-909f-2eedb8666c24)

kseniyashaydurova replied:
Pls implement sending emails and time measurement calculation as a separate workflow steps

LePaVuki replied:
Implemented the feature. An example of the email:

![image](https://github.com/user-attachments/assets/c1c4eb86-542d-47d7-b32c-9354da247f1b)

kseniyashaydurova commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+# nginx:mainline-alpine3.20-slim
+FROM nginx@sha256:e9293c9bedb0db866e7d2b69e58131db4c2478e6cd216cdd99b134830703983a AS nginx-ang
```
You don't have a multi-stage build, so you don't need a `AS nginx-ang` stuff

LePaVuki replied:
Got rid of the stage name

kseniyashaydurova commented on angular/nginx.conf:
Pls don't touch this file and create your own file in the repo root that will allow to process requests coming to container

LePaVuki replied:
Reverted the changes of the original nginx.conf
Now the project uses modified config file, located in the base directory

kseniyashaydurova commented on compose.yaml:
It is a good practise to name this file `docker-compose.yaml` usually

LePaVuki replied:
Renamed the file

kseniyashaydurova reviewed (changes requested):
Cool! It really works!
![image](https://github.com/user-attachments/assets/7e32e863-924d-42a6-8f61-60ebb3856890)

But pls address the PR feedback and:
1. Pls add test task description to PR summary
2. Missed steps related to sending emails in success and error cases
3. Can you do bonus step too?
4. See my feedback in comments

LePaVuki committed:
Got rid of staging in the dockerfile

LePaVuki committed:
Renamed compose.yaml -> docker-compose.yaml

LePaVuki committed:
Changed steps names to lowercase

LePaVuki committed:
Changed main nginx.conf file

Created a copy of nginx.conf in the root folder and reverted the old file content to initial

LePaVuki committed:
Removed unnecesary trigger to run action on pull-request

LePaVuki committed:
Changed working directory to angular for tests

Running tests and installation steps in the angular/ directory.
Added Google Chrome initialisation for Karma tests

LePaVuki committed:
Got rid of permissions, swapped to native npm build

LePaVuki committed:
Changed workflow to use github.token instead of secrets.GITHUB_TOKEN

LePaVuki committed:
Got rid of the browser setup in the workflow

LePaVuki committed:
Added xvfb action to the Karma Tests step

LePaVuki committed:
Fixed xvfb action in ci.yml

LePaVuki reviewed (commented):
By following your instructions, I altered the project, as requested. I have a few questions left after the work.

---
## Questions:


1. By default GitHub has the option to notify the initiator of an Action via email. Email contains the time it took an Action to finish and the type of the outcome. Is that enough for a project, or do I need to set up additional steps inside the workflow with actions from the Marketplace?

2. After changing working directories to ./angular, the tests on the remote GitHub machine are failing, due to Karma (testing framework) not being able to launch web browser correctly. The issue is in the lack of display for the GUI. At least that's what the debug message says. But I explicitly specified to launch the tests with "ChromeHeadless", which should remove the need for a display. Do I need to generate and modify configs for Karma, to overcome this issue? Or the solution is in additional steps in the workflow?

---
I hope I'm not asking too much, but those are the main problems I faced and struggling to solve. And thank you for your feedback

---
Mentions:
@kseniyashaydurova 
@dmitry-mightydevops

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,64 @@
+    - name: Checkout
+      uses: actions/checkout@v3
+
+    # - name: Setup Chrome
+    #   id: setup-chrome
+    #   uses: browser-actions/setup-chrome@v1
+    #   with:
+    #     install-dependencies: true
```
don't keep commented code in the pipeline - it makes reading much harder

LePaVuki replied:
Actually I wanted to ask about necessity to include such step in the workflow. But your advise on how to run Karma tests helped to resolve my issue

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,64 @@
+    - name: Install Angular
+      run: npm ci
+
+    - name: Run Karma Tests
+      uses: coactions/setup-xvfb@v1.0.1
+      with:
+        run: npm test --no-watch --no-progress --browsers=ChromeHeadlessCI
+      # env:
+      #   CHROME_BIN: ${{ steps.setup-chrome.outputs.chrome-path }}
```
try with
```
- name: Run Karma Tests
  id: run-tests
  run: npm run test -- --watch=false --browsers=ChromeHeadless 
```

LePaVuki replied:
Thank you for your advise, it helped

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+# nginx:mainline-alpine3.20-slim
+FROM nginx@sha256:e9293c9bedb0db866e7d2b69e58131db4c2478e6cd216cdd99b134830703983a
+
+COPY ./nginx.conf /etc/nginx/nginx.conf
+COPY ./angular/dist/angular-starter /var/www/html
+
+RUN chown -R nginx:nginx /var/cache/nginx && \
+        chown -R nginx:nginx /var/log/nginx && \
+        chown -R nginx:nginx /etc/nginx/conf.d
+RUN touch /var/run/nginx.pid && \
+        chown -R nginx:nginx /var/run/nginx.pid
+
+USER nginx
```
your custom dockerfile should explicitly say what CMD to run
don't rely on parent image default cmd. What if it has been compromised and CMD was modified with some malicious code which you will trust 100%.

LePaVuki replied:
Added explicit CMD at the end of my Dockerfile to run Nginx

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,35 @@
+  include /etc/nginx/mime.types;
+  default_type application/octet-stream;
+
+  ssl_protocols TLSv1 TLSv1.1 TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
+  ssl_prefer_server_ciphers on;
```
when you enable options
- you need to know the impact
- why did you enable SSL protocols where do you don't have SSL cert for the sever/location below?

LePaVuki replied:
Those options were there in the default config of the forked repo. The specification on my task described the actions I had to take a little vague. Including files to the final image, among them, I assumed, had to be nginx.conf. Initially I thought I had to make adjustments to it, not creating the conf file from scratch. And therefore I left those lines of code intact.

I cleaned up my current config file, to leave only necessary settings for the server.

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,35 @@
+}
+
+http {
+  sendfile on;
+  tcp_nopush on;
+  types_hash_max_size 2048;
```
what problem are you trying to solve?

LePaVuki replied:
Cleaned up the config file, same reason as above

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,35 @@
+
+  access_log /var/log/nginx/access.log;
+  gzip on;
+  include /etc/nginx/conf.d/*.conf;
```
why do you include? What if there is a site in default image that will get executed or some settings that will affect your environment?

LePaVuki replied:
Cleaned up the config file, same reason as above

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,35 @@
+  server {
+    listen 8080;
+
+    index index.html index.htm index.nginx-debian.html;
```
stick with just one, that your build produces. No need to solve million of use cases.

LePaVuki replied:
Left only index.html

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,35 @@
+    index index.html index.htm index.nginx-debian.html;
+    server_name _;
+    location / {
+      root /var/www/html/browser;
+      try_files $uri $uri/ /index.html;
```
this is not safe, you allow render any file passed as $uri, so if you have something like .env with secret values inside - it will be rendered as open text.

learn how to use this
https://nginx.org/en/docs/http/ngx_http_core_module.html#try_files

google has these tutorials and everyone is copying without analyzing what exactly they had done. And the impact, that will cause
- security breaches
- downtime

When you have a configuration, it's not enough that it **worked** you need to know how/why it has worked. Otherwise you will become a chatgpt victim

find a better way

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,64 @@
+
+jobs:
+
+  create-and-push-image:
```
```suggestion
  build:
```

dmitry-mightydevops replied:
use short semantic names when possible

LePaVuki replied:
shortened the name

dmitry-mightydevops reviewed (changes requested):
here you go

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,64 @@
+        working-directory: ./angular
+
+    steps:
+    - name: Checkout
```
Where is the step to setup corresponding node version?

P.S. node version should be taken from package.json file

LePaVuki replied:
Added the step with Node, also made adjustments to package.json, to include the Node version

kseniyashaydurova commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+COPY ./nginx.conf /etc/nginx/nginx.conf
+COPY ./angular/dist/angular-starter /var/www/html
+
+RUN chown -R nginx:nginx /var/cache/nginx && \
```
1. Can you pls check whether all these operations are needed?
2. Try to pass multiple files to single `chown` command

LePaVuki replied:
Those commands were used by me to solve the problem of running Nginx as non root user.
I decided to switch to the other option: pull an official unprivileged Nginx image. It seems more suiting for the task
Since now I found a simpler solution, chown command not needed anymore

kseniyashaydurova commented on docker-compose.yaml:
```diff
@@ -0,0 +1,7 @@
+services:
+
+  nginx-ang:
```
Bad name, not clear from the first sight. It is better to use smth sensible and clear, like `nginx-angular`

LePaVuki replied:
Changed name to suggested

kseniyashaydurova commented on docker-compose.yaml:
```diff
@@ -0,0 +1,7 @@
+services:
+
+  nginx-ang:
+    container_name: nginx-ang
```
The same as above

kseniyashaydurova commented on docker-compose.yaml:
```diff
@@ -0,0 +1,7 @@
+
+  nginx-ang:
+    container_name: nginx-ang
+    image: "ghcr.io/lepavuki/angular-react-devops-case:master"
```
You can omit `"` here

kseniyashaydurova commented on nginx.conf:
```diff
@@ -0,0 +1,35 @@
+worker_processes auto;
```
Let's specify user under that nginx processes will work - https://nginx.org/en/docs/ngx_core_module.html#user

LePaVuki replied:
Do I still need to specify the user, if I'm using unprivileged image of Nginx? I switch to user nginx in the dockerfile, before executing CMD nginx

kseniyashaydurova replied:
Yes, potentially we can run docker container as root user if we need so, even if it has an `nginx` default user, in such a case nginx with this config would be started by `root` user instead of `nginx` one

kseniyashaydurova reviewed (changes requested):
Let's have at least one successful workflow call at the end
![image](https://github.com/user-attachments/assets/5e47e060-5bdf-4eb3-9eda-3098397e787d)

kseniyashaydurova commented:
> By default GitHub has the option to notify the initiator of an Action via email. Email contains the time it took an Action to finish and the type of the outcome. Is that enough for a project, or do I need to set up additional steps inside the workflow with actions from the Marketplace?

Pls implement your own steps to calculate workflow execution time and send emails.

> After changing working directories to ./angular, the tests on the remote GitHub machine are failing, due to Karma (testing framework) not being able to launch web browser correctly. The issue is in the lack of display for the GUI. At least that's what the debug message says. But I explicitly specified to launch the tests with "ChromeHeadless", which should remove the need for a display. Do I need to generate and modify configs for Karma, to overcome this issue? Or the solution is in additional steps in the workflow?

No, no need to change Karma configs. You can run tests like this: `npm run coverage -- --browsers=ChromeHeadless`

LePaVuki committed:
Minor renames, setup Node step

LePaVuki committed:
Changed Karma tests run

LePaVuki committed:
Changed docker-compose to pull correct image name

LePaVuki committed:
Simplified nginx.conf

LePaVuki committed:
Adjusted nginx.conf

LePaVuki committed:
Specified node version in package.json

LePaVuki committed:
Added email notification and time tracking

LePaVuki committed:
Reverted docker-compose from local building to web image pulling

LePaVuki committed:
Specified CMD in dockerfile to run Nginx

LePaVuki committed:
Added mime types to include in the Nginx.conf

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+      if: always()
+      uses: dawidd6/action-send-mail@v3
+      with:
+        server_address: smtp.gmail.com
```
This shouldn't be hardcoded at steps level as well as `server_port`, move it to inputs - https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callinputs

Your codes should be always parametrized, so they could be reused for different cases

LePaVuki replied:
Moved to inputs in the build-angular.yml. Now it can be set dynamically

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+      with:
+        server_address: smtp.gmail.com
+        server_port: 465
+        username: ${{secrets.email_username}}
```
Pls specify secrets in `workflow_call` too - https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callsecrets

LePaVuki replied:
In the inputs now

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+        from: ${{ secrets.mail_sender }}
+        body: Build job of ${{github.repository}} completed successfully!
+        html_body: |
+          # Job on ${{ github.repository }}: ${{ job.status }}
```
Plos show screenshot of failed job too

LePaVuki replied:
added screenshots of both outcomes

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+          # Job on ${{ github.repository }}: ${{ job.status }}
+          ---
+          Time of execution: ${{ steps.duration.outputs.duration }}
+        convert_markdown: true
```
Usually it is better to locate one-liner arguments above multilines for better readability

LePaVuki replied:
moved the line higher

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+    branches: [ "master" ]
+
+env:
+  REGISTRY: ghcr.io
```
Let's configure it also as `input` - https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callinputs

LePaVuki replied:
Moved to inputs in the build-angular.yml

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+
+env:
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ${{ github.repository }}
```
Let's move this to inputs too

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+        push: true
+        tags: ${{ steps.meta.outputs.tags }}
+        labels: ${{ steps.meta.outputs.labels }}
+
```
Pls remove redundant line here to have the same formatting everywhere

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,87 @@
+        subject: GH Action ${{ github.repository }} ${{ github.workflow }} ${{ job.status }}
+        to: ${{ secrets.mail_receiver }}
+        from: ${{ secrets.mail_sender }}
+        body: Build job of ${{github.repository}} completed successfully!
```
If you job failed, you will get a `completed successfully` body?

LePaVuki replied:
Fixed the redundant line. Now it's displaying result dynamically, fully in html-body

kseniyashaydurova commented on .github/workflows/ci.yml:
Pls show emails screenshots for successful and failed jobs both

LePaVuki replied:
![image](https://github.com/user-attachments/assets/3147ca1f-859f-4e60-8d3b-8564600a66c7)
![image](https://github.com/user-attachments/assets/e227c55f-436e-4351-ba34-745c301ebfa6)

kseniyashaydurova commented on docker-compose.yaml:
```diff
@@ -0,0 +1,7 @@
+
+  nginx-angular:
+    container_name: nginx-angular
+    image: ghcr.io/lepavuki/angular-react-devops-case:pr-3
```
Pls update your workflow to produce nice tags for docker images, i.e. should be at least `latest` one

LePaVuki replied:
Added "latest" tag handling

kseniyashaydurova reviewed (changes requested):
Needs a bit more changes

LePaVuki committed:
Resructurised workflows

LePaVuki committed:
Separated most of the jobs to separate workflows

LePaVuki committed:
Changed success display of jobs in the email

LePaVuki committed:
got rid of separate time count

LePaVuki committed:
Simplified workflow to 2 separate files

LePaVuki committed:
changed nginx.conf to incorrect for workflow test

LePaVuki committed:
Update ci.yml

LePaVuki committed:
fixed nginx.conf and build-angular

LePaVuki committed:
broke Build Angular step to test workflow

LePaVuki committed:
Changed tagging of the docker image

LePaVuki committed:
Update readability of build-angular.yml

dmitry-mightydevops commented on .github/workflows/build-angular.yml:
```diff
@@ -0,0 +1,125 @@
+name: Build Angular
```
```suggestion
name: build-angular
```
match the name with the file name

dmitry-mightydevops commented on .github/workflows/build-angular.yml:
your implementation needs to have a proper documentation at the top, see this as an example

```yaml
# --------------------------------------------------------------
# GitHub Actions Workflow: <Workflow Name>
# 
# Description:
#   This workflow automates the process of [describe the primary task 
#   or purpose, e.g., "building and testing the project," "deploying 
#   the application to production," etc.]. It consists of multiple 
#   jobs with various steps to ensure [mention any goals like "code 
#   quality," "successful deployment," or "artifact generation"].
#
# Triggers:
#   - on: 
#       - push: Executes on a push to [specific branches or paths if any].
#       - pull_request: Executes on a pull request to [specific branches].
#       - schedule: Runs on a cron schedule for [mention frequency, 
#         e.g., "daily builds" or "weekly maintenance"].
#
# Jobs:
#   - <Job Name 1>: [Brief description of the job, e.g., "Sets up 
#       the environment and installs dependencies."]
#   - <Job Name 2>: [e.g., "Runs unit and integration tests to 
#       validate changes."]
#   - <Job Name 3>: [e.g., "Deploys to the production environment 
#       upon successful completion of all tests."]
#
# Notes:
#   - Requires the following secrets to be configured in the GitHub 
#     repository: [list secrets, e.g., "AWS_ACCESS_KEY", "DOCKER_TOKEN"].
#   - [Any prerequisites or dependencies, e.g., "Ensure Docker is 
#     installed on self-hosted runners."]
#   - [Any environment configurations or versions required, e.g., 
#     "Node.js 14+" or "Python 3.9"]
#
# Expected Outcome:
#   - Successful execution of this workflow should result in [state 
#     the outcome, e.g., "a new Docker image published to Docker Hub" 
#     or "a deployed app on production servers"]. 
#
# --------------------------------------------------------------

```

dmitry-mightydevops commented on .github/workflows/build-angular.yml:
```diff
@@ -0,0 +1,125 @@
+name: Build Angular
+
+on:
+  workflow_call:
+
+    inputs:
+
+      mail_address:
+        description: "Mail server address"
+        default: "smtp.gmail.com"
+        required: true
+        type: string
+
+      mail_port:
+        description: "Mail server port"
```
make condensed code and remove empty lines.

dmitry-mightydevops commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,24 @@
+name: CI pipeline
```
```suggestion
name: ci
```

LePaVuki committed:
Renamed workflow to fit with the code

Co-authored-by: Dmitry Semenov <70822710+dmitry-mightydevops@users.noreply.github.com>

LePaVuki committed:
Rename of the workflow

Co-authored-by: Dmitry Semenov <70822710+dmitry-mightydevops@users.noreply.github.com>

LePaVuki committed:
Condensed code

LePaVuki committed:
Added the workflows description
